larky/src/main/resources/vendor/ISO8583Decoder.star

```diff
@@ -0,0 +1,505 @@
+    # and never enters the loop to create the variable.
+    # Set `field_key` to the last mandatory one: primary bitmap.
+    field_key = "p"
+
+    for field_key in [str(i) for i in sorted(fields)]:
+        # Secondary bitmap is already decoded in _decode_bitmaps
+        if field_key == "1":
+            continue
+        idx = _decode_field(s, doc_dec, doc_enc, idx, field_key, spec)
+
+    if idx != len(s):
+        fail(" DecodeError(\n            \"Extra data after last field\", s, doc_dec, doc_enc, idx, field_key\n        )")
+
+    return doc_dec, doc_enc
+
+
+#
+# Private interface
+#
+
+
+def _decode_header(
+    s,
+    doc_dec,
+    doc_enc,
+    idx,
+    spec,
+):
+    r"""Decode ISO8583 header data if present.
+
+    Parameters
+    ----------
+    s : bytes or bytearray
+        Encoded ISO8583 data
+    doc_dec : dict
+        Dict containing decoded ISO8583 data
+    doc_enc : dict
+        Dict containing encoded ISO8583 data
+    idx : int
+        Current index in ISO8583 byte array
+    spec : dict
+        A Python dict defining ISO8583 specification.
+        See :mod:`iso8583.specs` module for examples.
+
+    Returns
+    -------
+    int
+        Index in ISO8583 byte array where parsing of the header ended
+
+    Raises
+    ------
+    DecodeError
+        An error decoding ISO8583 bytearray.
+    """
+
+    # Header is not expected according to specifications
+    if spec["h"]["max_len"] <= 0:
+        return idx
+
+    return _decode_field(s, doc_dec, doc_enc, idx, "h", spec)
+
+
+def _decode_type(
+    s,
+    doc_dec,
+    doc_enc,
+    idx,
+    spec,
+) :
+    r"""Decode ISO8583 message type.
+
+    Parameters
+    ----------
+    s : bytes or bytearray
+        Encoded ISO8583 data
+    doc_dec : dict
+        Dict containing decoded ISO8583 data
+    doc_enc : dict
+        Dict containing encoded ISO8583 data
+    idx : int
+        Current index in ISO8583 byte array
+    spec : dict
+        A Python dict defining ISO8583 specification.
+        See :mod:`iso8583.specs` module for examples.
+
+    Returns
+    -------
+    int
+        Index in ISO8583 byte array where parsing of message type ended
+
+    Raises
+    ------
+    DecodeError
+        An error decoding ISO8583 bytearray.
+    """
+
+    # Message type is a set length in ISO8583
+    if spec["t"]["data_enc"] == "b":
+        f_len = 2
+    else:
+        f_len = 4
+
+    doc_dec["t"] = ""
+    doc_enc["t"] = {"len": bytes(r"", encoding='utf-8'), "data": bytes(s[idx : idx + f_len])}
+
+    if len(s[idx : idx + f_len]) != f_len:
+        fail(" DecodeError(\n            f\"Field data is {len(s[idx:idx + f_len])} bytes, expecting {f_len}\",\n            s,\n            doc_dec,\n            doc_enc,\n            idx,\n            \"t\",\n        )")
+
+    if spec["t"]["data_enc"] == "b":
+        doc_dec["t"] = s[idx : idx + f_len].hex().upper()
+    else:
+        doc_dec["t"] = s[idx : idx + f_len].decode(spec["t"]["data_enc"])
```

Comment:
we do not support `.decode`, you can use:

https://github.com/verygoodsecurity/starlarky/blob/b84752769ea674180ff65d772db69aad379948ba/larky/src/test/resources/vendor_tests/test_escapes.star#L6

and 
https://github.com/verygoodsecurity/starlarky/blob/b84752769ea674180ff65d772db69aad379948ba/larky/src/test/resources/vendor_tests/test_escapes.star#L16

but obviously replace the encoding with `spec["t"]["data_enc"]`